File: backend/src/main/java/com/multi/finance/entity/Bill.java
package com.multi.finance.entity;

import com.multi.finance.enums.BillSource;
import com.multi.finance.enums.BillStatus;
import com.multi.finance.enums.BillType;
import com.multi.finance.enums.BusinessType;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.Fetch;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "bills")
public class Bill {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "bill_number")
    private String billNumber;

    @Enumerated(EnumType.STRING)
    @Column(name = "business", nullable = false)
    private BusinessType business;

    @Column(name = "division", nullable = false)
    private String division;

    @Enumerated(EnumType.STRING)
    @Column(name = "bill_source", nullable = false)
    private BillSource billSource;

    @Enumerated(EnumType.STRING)
    @Column(name = "bill_type", nullable = false)
    private BillType billType;

    @Column(name = "customer_name", nullable = false)
    private String customerName;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "customer_id")
    private Customer customer;

    @Column(name = "total_amount", nullable = false)
    private BigDecimal totalAmount;

    @Column(name = "area")
    private String area;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private BillStatus status;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "current_holder_id")
    private Worker currentHolder;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "entered_by", nullable = false)
    private User enteredBy;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "received_by")
    private User receivedBy;

    @Column(name = "received_at")
    private LocalDateTime receivedAt;

    @Column(name = "bill_date" ,nullable = false)
    private LocalDate billDate;

    private String notes;

    @Column(name = "amount_paid", nullable = false)
    private BigDecimal amountPaid;

    @Column(name = "balance_remaining", nullable = false)
    private BigDecimal balanceRemaining;

    /**
     * Everything the customer has sent back on this bill, damage and salable together.
     * Held apart from {@link #totalAmount} so the invoiced figure survives a return and
     * an admin reversal is a recompute rather than a compensating entry.
     * Always go through {@code BillBalance} rather than reading this directly.
     */
    @Column(name = "returns_total", nullable = false)
    @Builder.Default
    private BigDecimal returnsTotal = BigDecimal.ZERO;

    @Column(name = "fully_paid", nullable = false)
    private Boolean fullyPaid;

    /** Manually marked as collection-only — skip delivery classification regardless of bill age. */
    @Column(name = "collection_only", nullable = false)
    @Builder.Default
    private Boolean collectionOnly = false;

    /**
     * True for SYSTEM bills that will be reconciled via End-of-Month linking to DRAFT/MANUAL bills.
     * When true: excluded from Summary Load + Stock Item Entry; only appears in Linking tab.
     * Stock is covered by the linked children — no separate stock movement needed.
     */
    @Column(name = "will_be_linked", nullable = false)
    @Builder.Default
    private Boolean willBeLinked = false;

    /** Set by admin once the system linking bill's qty matches all linked children — final sign-off. */
    @Column(name = "stock_reconciled", nullable = false)
    @Builder.Default
    private Boolean stockReconciled = false;

    /** Set by admin once the savings amount (children total − system amount) has been collected. */
    @Column(name = "savings_collected", nullable = false)
    @Builder.Default
    private Boolean savingsCollected = false;

    @Column(name = "stock_cleared", nullable = false)
    @Builder.Default
    private Boolean stockCleared = false;

    /**
     * How the goods reached the customer. UNSPECIFIED on everything entered before
     * deliveries were recorded — left as it is rather than guessed into a mode that
     * would then be counted as fact.
     */
    @Enumerated(EnumType.STRING)
    @Column(name = "delivery_mode", nullable = false)
    @Builder.Default
    private com.multi.finance.enums.DeliveryMode deliveryMode =
            com.multi.finance.enums.DeliveryMode.UNSPECIFIED;

    /** The lorry round this went out on. Set only when the mode is ROUTE. */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "delivery_run_id")
    private DeliveryRun deliveryRun;

    /**
     * The bill this one's money is actually collected on — a hand-written bill for the
     * same sale.
     *
     * <p>This record is real and keeps its stock; it simply is not the one being paid.
     * So it stops counting as outstanding, leaves the aging report, and closes when the
     * bill it points at is paid off.
     *
     * <p>Not the same as bill_stock_links, which ties one system bill to many manual
     * ones for month-end stock. This is one-to-one and about the money.
     */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "settled_on_bill_id")
    private Bill settledOn;

    @Column(name = "settled_on_at")
    private LocalDateTime settledOnAt;

    @Column(name = "settled_on_by", length = 100)
    private String settledOnBy;

    @Column(name = "settled_on_note", length = 300)
    private String settledOnNote;

    /** True when the money for this bill is collected on another one. */
    public boolean isSettledElsewhere() {
        return settledOn != null;
    }

    /** Why this bill was cancelled. Required when cancelling. */
    @Column(name = "cancel_reason", length = 300)
    private String cancelReason;

    @Column(name = "cancelled_by", length = 100)
    private String cancelledBy;

    @Column(name = "cancelled_at")
    private LocalDateTime cancelledAt;

    /**
     * Kept off the aging report. The balance is still owed and still on the bill —
     * this only says it is not chaseable debt worth reporting, so the report does not
     * overstate what can actually be collected.
     */
    @Column(name = "excluded_from_aging", nullable = false)
    @Builder.Default
    private Boolean excludedFromAging = false;

    @Column(name = "aging_exclusion_reason", length = 300)
    private String agingExclusionReason;

    @Column(name = "aging_excluded_by", length = 100)
    private String agingExcludedBy;

    @Column(name = "aging_excluded_at")
    private LocalDateTime agingExcludedAt;

    @Column(name = "created_at")
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
}

File: finance-frontend/src/app/core/services/bill.ts
import { Injectable } from '@angular/core';
import { environment } from '../../../environments/environment';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';


export interface AgingCustomerEntry {
  customerName: string;
  customerId: number | null;
  area: string | null;
  totalOutstanding: number;
  overdue: number;
  current: number;
  days31to60: number;
  days61to90: number;
  days91plus: number;
  billCount: number;
  oldestBillDate: string;
  lastPaymentDate: string | null;
  cashPending: number;
  cashFollowUp: number;
  cashUrgent: number;
  cashSerious: number;
}

export interface AgingAreaSummary {
  area: string;
  totalOutstanding: number;
  overdue: number;
  current: number;
  days31to60: number;
  days61to90: number;
  days91plus: number;
  cashPending: number;
  cashSerious: number;
  customerCount: number;
  billCount: number;
  customers: AgingCustomerEntry[];
}

export interface AgingReportResponse {
  grandTotalOutstanding: number;
  grandOverdue: number;
  grandCashPending: number;
  grandCashSerious: number;
  totalCustomers: number;
  totalBills: number;
  topCustomers: AgingCustomerEntry[];
  allCustomers: AgingCustomerEntry[];
  byArea: AgingAreaSummary[];
}

export interface BillSequenceGap {
  billSource: string;
  totalBills: number;
  firstNumber: number;
  lastNumber: number;
  missingCount: number;
  missingNumbers: string[];
}

export interface BillFilter {
  business?: string;
  status?: string;
  excludeCompleted?: boolean;
  from?: string;
  to?: string;
}

export interface BillResponse {
  /**
   * Collected on the hand-written bill for the same sale. This bill is real and its
   * stock went out; it is simply not the one being paid, so it leaves the aging report
   * and closes when that bill is paid off.
   */
  settledOnBillId?: number;
  settledOnBillNumber?: string;
  settledOnStatus?: string;
  settledOnNote?: string;
  settledOnBy?: string;

  /** Why it was voided. */
  cancelReason?: string;
  cancelledBy?: string;
  cancelledAt?: string;

  /** How it reached the customer, and the lorry round it travelled on. */
  deliveryMode?: 'UNSPECIFIED' | 'ROUTE' | 'IMMEDIATE' | 'STORE_PICKUP';
  deliveryRunId?: number;
  deliveryRunArea?: string;
  deliveryRunDate?: string;

  excludedFromAging?: boolean;
  agingExclusionReason?: string;
  agingExcludedBy?: string;
  agingExcludedAt?: string;

  id: number;
  billNumber: string;
  business: string;
  division: string;
  billType: string;
  billSource: string;
  customerName: string;
  /** Null on bills whose customer was only ever typed as a name. */
  customerId?: number;
  totalAmount: number;
  amountPaid: number;
  balanceRemaining: number;
  fullyPaid: boolean;
  status: string;
  workerId: number | null;
  workerName: string | null;
  enteredByName: string;
  receivedByName: string | null;
  receivedAt: string | null;
  area: string | null;
  billDate: string;
  notes: string | null;
  createdAt: string;
  willBeLinked?: boolean;
  stockCleared?: boolean;
  collectionOnly?: boolean;
}

export interface SkipReviewResponse {
  id: number;
  business: string;
  skippedBillNumber: string;
  relatedBillNumber: string | null;
  customerName: string | null;
  submittedByName: string | null;
  submittedAt: string;
}

/** A number offered in the create-bill dropdown. */
export interface BillNumberOption {
  number: number;
  /** Sits in a hole below the highest number used — never entered, and not an approved skip. */
  missing: boolean;
}

@Injectable({
  providedIn: 'root',
})
export class Bill {
  private apiUrl = `${environment.apiUrl}/bills`;

  constructor(private http: HttpClient) {}

  globalSearch(q: string): Observable<BillResponse[]> {
    return this.http.get<BillResponse[]>(`${this.apiUrl}/search`, { params: { q } });
  }

  getNextBillNumbers(business: string, billSource: string): Observable<BillNumberOption[]> {
    return this.http.get<BillNumberOption[]>(`${this.apiUrl}/next-numbers`,
      { params: { business, billSource } });
  }

  getPendingSkips(): Observable<SkipReviewResponse[]> {
    return this.http.get<SkipReviewResponse[]>(`${this.apiUrl}/skip-reviews`);
  }

  approveSkip(id: number): Observable<void> {
    return this.http.patch<void>(`${this.apiUrl}/skip-reviews/${id}/approve`, {});
  }

  rejectSkip(id: number): Observable<void> {
    return this.http.patch<void>(`${this.apiUrl}/skip-reviews/${id}/reject`, {});
  }

  getBills(filter?: BillFilter): Observable<BillResponse[]> {
    let params = new HttpParams();
    if (filter?.business) params = params.set('business', filter.business);
    if (filter?.status)   params = params.set('status', filter.status);
    if (filter?.excludeCompleted !== undefined) params = params.set('excludeCompleted', String(filter.excludeCompleted));
    if (filter?.from) params = params.set('from', filter.from);
    if (filter?.to)   params = params.set('to', filter.to);
    return this.http.get<BillResponse[]>(this.apiUrl, { params });
  }

  getLinkingBills(): Observable<BillResponse[]> {
    return this.http.get<BillResponse[]>(`${this.apiUrl}/linking`);
  }

  getOverdueCount(): Observable<{ count: number }> {
    return this.http.get<{ count: number }>(`${this.apiUrl}/overdue-count`);
  }

  getBillById(id: number): Observable<BillResponse> {
    return this.http.get<BillResponse>(`${this.apiUrl}/${id}`);
  }

  createBill(payload: any): Observable<BillResponse> {
    return this.http.post<BillResponse>(this.apiUrl, payload);
  }

  assignBill(id: number, workerId: number): Observable<BillResponse> {
    return this.http.patch<BillResponse>(`${this.apiUrl}/${id}/assign`, { workerId });
  }

  bulkAssignBills(billIds: number[], workerId: number): Observable<BillResponse[]> {
    return this.http.patch<BillResponse[]>(`${this.apiUrl}/bulk-assign`, { billIds, workerId });
  }

  bulkMarkReceived(billIds: number[]): Observable<BillResponse[]> {
    return this.http.patch<BillResponse[]>(`${this.apiUrl}/bulk-receive`, { billIds });
  }

  bulkMarkShopReceived(billIds: number[]): Observable<BillResponse[]> {
    return this.http.patch<BillResponse[]>(`${this.apiUrl}/bulk-shop-receive`, { billIds });
  }

  markReceived(id: number): Observable<BillResponse> {
    return this.http.patch<BillResponse>(`${this.apiUrl}/${id}/receive`, {});
  }

  updateBill(id: number, payload: any): Observable<BillResponse> {
    return this.http.put<BillResponse>(`${this.apiUrl}/${id}`, payload);
  }

  markShopReceived(id: number): Observable<BillResponse> {
    return this.http.patch<BillResponse>(`${this.apiUrl}/${id}/shop-receive`, {});
  }

  markStockCleared(id: number): Observable<BillResponse> {
    return this.http.patch<BillResponse>(`${this.apiUrl}/${id}/mark-stock-cleared`, {});
  }

  markCompleted(id: number): Observable<BillResponse> {
    return this.http.patch<BillResponse>(`${this.apiUrl}/${id}/complete`, {});
  }

  /**
   * Voids a bill. The reason is required: the bill keeps its number and stays in the
   * run forever, and whoever finds it later has only this to explain it.
   */
  cancelBill(id: number, reason: string): Observable<BillResponse> {
    return this.http.patch<BillResponse>(`${this.apiUrl}/${id}/cancel`, { reason });
  }

  /**
   * Records that this bill's money is collected on a hand-written one instead.
   *
   * <p>Used where a bill used to be cancelled as a duplicate. Cancelling said the sale
   * never happened; it did, and the stock went out on it. Linking keeps the record and
   * only stops the chasing.
   */
  linkSettlement(id: number, targetBillId: number, note?: string): Observable<BillResponse> {
    return this.http.patch<BillResponse>(`${this.apiUrl}/${id}/settle-on`, {
      targetBillId,
      note: note ?? '',
    });
  }

  unlinkSettlement(id: number): Observable<BillResponse> {
    return this.http.delete<BillResponse>(`${this.apiUrl}/${id}/settle-on`);
  }

  /** Manual bills this one could be collected on — same customer, within a month. */
  getSettleCandidates(id: number): Observable<BillResponse[]> {
    return this.http.get<BillResponse[]>(`${this.apiUrl}/${id}/settle-candidates`);
  }

  /** The bills collected on this one. */
  getSettledByBills(id: number): Observable<BillResponse[]> {
    return this.http.get<BillResponse[]>(`${this.apiUrl}/${id}/settled-by`);
  }

  toggleCollectionOnly(id: number): Observable<void> {
    return this.http.patch<void>(`${this.apiUrl}/${id}/toggle-collection-only`, {});
  }

  deleteBill(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  getAgingReport(business: string = 'RAINCO'): Observable<AgingReportResponse> {
    return this.http.get<AgingReportResponse>(`${this.apiUrl}/aging-report`, {
      params: { business },
    });
  }

  /** Printable aging data — area and billType optional (omit = all). */
  getAgingExport(business: string, area?: string, billType?: string, sort?: string): Observable<AgingExport> {
    let params = new HttpParams().set('business', business);
    if (area) params = params.set('area', area);
    if (billType) params = params.set('billType', billType);
    if (sort) params = params.set('sort', sort);
    return this.http.get<AgingExport>(`${this.apiUrl}/aging-report/export`, { params });
  }

  downloadAgingExcel(business: string, area?: string, billType?: string, sort?: string): Observable<Blob> {
    let params = new HttpParams().set('business', business);
    if (area) params = params.set('area', area);
    if (billType) params = params.set('billType', billType);
    if (sort) params = params.set('sort', sort);
    return this.http.get(`${this.apiUrl}/aging-report/export.xlsx`, {
      params, responseType: 'blob',
    });
  }

  findSequenceGaps(business: string = 'RAINCO'): Observable<BillSequenceGap[]> {
    return this.http.get<BillSequenceGap[]>(`${this.apiUrl}/sequence-gaps`, {
      params: { business },
    });
  }

  // ── Bill Review ────────────────────────────────────────────────────────────

  getUnreviewedBills(): Observable<BillResponse[]> {
    return this.http.get<BillResponse[]>(`${this.apiUrl}/review/unreviewed`);
  }

  getUnreviewedCount(): Observable<{ count: number }> {
    return this.http.get<{ count: number }>(`${this.apiUrl}/review/unreviewed-count`);
  }

  markBillsReviewed(billIds: number[]): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/review/mark`, { billIds });
  }

  markAllBillsReviewed(): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/review/mark-all`, {});
  }

  /**
   * Hides a bill from the aging report, or puts it back. Admin only. Never a delete —
   * the balance stays owed, it just stops being reported as chaseable.
   */
  setAgingVisibility(billId: number, excluded: boolean, reason?: string): Observable<BillResponse> {
    return this.http.patch<BillResponse>(
      `${this.apiUrl}/${billId}/aging-visibility`, { excluded, reason });
  }

  /** Bills currently hidden, so an exclusion cannot be quietly forgotten. */
  getAgingExcluded(business: string): Observable<BillResponse[]> {
    return this.http.get<BillResponse[]>(`${this.apiUrl}/aging-report/excluded`,
      { params: new HttpParams().set('business', business) });
  }
}

export interface AgingExportBillRow {
  billNumber: string;
  billDate: string;
  ageDays: number;
  customerName: string;
  area: string | null;
  billType: string;
  totalAmount: number;
  amountPaid: number;
  balance: number;
  bucket: string;
  workerName: string | null;
  lastPaymentDate: string | null;
}

export interface AgingExportCustomer {
  customerName: string;
  area: string | null;
  totalOutstanding: number;
  overdue: number;
  current: number;
  days31to60: number;
  days61to90: number;
  days91plus: number;
  cashPending: number;
  cashFollowUp: number;
  cashUrgent: number;
  cashSerious: number;
  billCount: number;
  oldestBillDate: string | null;
  lastPaymentDate: string | null;
}

export interface AgingExport {
  business: string;
  area: string | null;
  billType: string | null;
  generatedOn: string;
  creditCustomers: AgingExportCustomer[];
  cashCustomers: AgingExportCustomer[];
  bills: AgingExportBillRow[];
  totalCredit: number;
  totalCash: number;
  totalOutstanding: number;
  customerCount: number;
  billCount: number;

  /** Bills an admin kept off this report, and what they come to. */
  excludedCount?: number;
  excludedAmount?: number;
}
